test/hnsd-test.c

```diff
@@ -27,21 +41,328 @@ test_base32() {
+        hsk_dns_rrsig_rd_t *r = (hsk_dns_rrsig_rd_t *)rr->rd;
+        char name1[HSK_DNS_MAX_NAME_STRING] = {0};
+        assert(hsk_dns_name_to_string(r->signer_name, name1));
+        assert(strcmp(name1, record_read_vector.name1) == 0);
+        break;
+      }
+      case HSK_DNS_NSEC: {
+        hsk_dns_nsec_rd_t *r = (hsk_dns_nsec_rd_t *)rr->rd;
+        char name1[HSK_DNS_MAX_NAME_STRING] = {0};
+        assert(hsk_dns_name_to_string(r->next_domain, name1));
+        assert(strcmp(name1, record_read_vector.name1) == 0);
+        break;
+      }
+    }
+
+    hsk_dns_msg_free(msg);
+  }
+}
+
+void
+test_hsk_dns_msg_write(){
+  printf("  test_hsk_dns_msg_write\n");
+  for (int i = 0; i < ARRAY_SIZE(record_read_vectors_valid); i++) {
+    record_read_vector_t record_read_vector = record_read_vectors_valid[i];
+
+    // Build expected DNS message from test vector
+    size_t meta_len = 2 + 2 + 4 + 2; // type, class, ttl, rd size
+    size_t rd_len = record_read_vector.data_len;
+    size_t total_len = record_read_msg_len + meta_len + rd_len;
+    uint8_t data[total_len];
+    uint8_t *data_ = (uint8_t *)&data;
+
+    write_bytes(&data_, record_read_msg, record_read_msg_len);
+    write_u16be(&data_, record_read_vector.type); // type
+    write_u16be(&data_, HSK_DNS_IN);              // class
+    write_u32be(&data_, 0x00);                    // ttl
+    write_u16be(&data_, rd_len);                  // rd size
+    write_bytes(&data_, (uint8_t *)record_read_vector.data, rd_len);
+
+    // Build msg to serialize
+    hsk_dns_msg_t *msg = hsk_dns_msg_alloc();
+    hsk_dns_rr_t *rr = hsk_dns_rr_alloc();
+    hsk_dns_name_from_string(record_read_msg_qname, rr->name);
+    rr->type = record_read_vector.type;
+    rr->class = HSK_DNS_IN;
+    rr->ttl = 0;
+    rr->rd = hsk_dns_rd_alloc(record_read_vector.type);
+    hsk_dns_rrs_push(&msg->an, rr);
+
+    // Write names, convert from presentation format
+    switch (record_read_vector.type) {
+      case HSK_DNS_SOA:
+      case HSK_DNS_MX:
+      case HSK_DNS_SRV:
+      case HSK_DNS_RRSIG:
+      case HSK_DNS_NSEC:
+      case HSK_DNS_RP:
+        // Only testing simple records with just names
+        hsk_dns_msg_free(msg);
+        continue;
+      case HSK_DNS_NS: {
+        hsk_dns_ns_rd_t *ns = rr->rd;
+        hsk_dns_name_from_string(record_read_vector.name1, ns->ns);
+        break;
+      }
+      case HSK_DNS_CNAME: {
+        hsk_dns_cname_rd_t *cname = rr->rd;
+        hsk_dns_name_from_string(record_read_vector.name1, cname->target);
+        break;
+      }
+      case HSK_DNS_PTR: {
+        hsk_dns_ptr_rd_t *ptr = rr->rd;
+        hsk_dns_name_from_string(record_read_vector.name1, ptr->ptr);
+        break;
+      }
+      // case HSK_DNS_RP: {
+      //   hsk_dns_rp_rd_t *rp = rr->rd;
+      //   hsk_dns_name_from_string(record_read_vector.name1, rp->mbox);
+      //   hsk_dns_name_from_string(record_read_vector.name2, rp->txt);
+      //   break;
+      // }
```

Comment:
TODO, figured this out:

https://www.rfc-editor.org/rfc/rfc3597#section-4

>   To avoid such corruption, servers MUST NOT compress domain names
>    embedded in the RDATA of types that are class-specific or not well-
>    known.  This requirement was stated in [RFC1123] without defining the
>    term "well-known"; it is hereby specified that only the RR types
>    defined in [RFC1035] are to be considered "well-known".

RP is not considered "well known" since its not in 1035